consensus/src/experimental/buffer_manager.rs

```diff
@@ -257,17 +271,29 @@ impl StateManager {
             // buffer manager will stop
             self.end_epoch = true;
         } else {
-            // clear the buffer until (including) the ledger_info
-            while let Some(buffer_item) = self.buffer.pop_front() {
-                if buffer_item
-                    .get_commit_info()
-                    .match_ordered_only(ledger_info.commit_info())
+            let mut cursor = self.buffer.head.clone();
+            // look for the target ledger info:
+            // if found: we try to advance it to aggregated, if succeeded, we try persisting the items.
+            // if not found: it means the block is in BlockStore but not in the buffer
+            // either the block is just persisted, or has not been added to the buffer
+            // in either cases, we do nothing.
+            while cursor.is_some() {
                 {
-                    break;
+                    let buffer_item = take_elem(&cursor);
+                    let (attempted_item, aggregated, matching) =
+                        buffer_item.try_advance_to_aggregated_with_ledger_info(ledger_info.clone());
+                    set_elem(&cursor, attempted_item);
+                    if matching {
```

Comment:
if aggregated is enough?

Reply:
there are three cases:
1. not matched, not aggregated, we continue to the next iteration
2. matched, not aggregated, we end the loop since we already found the item
3. matched and aggregated, we call try_persisting to dequeue the items.

Reply:
I mean aggregated must be matched?

Reply:
yeah, but they share a break statement. So I put it inside matching.